packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/queue/queue.ts

```diff
@@ -0,0 +1,393 @@
+import * as subscription from '@aws-cdk/aws-sns-subscriptions';
+import * as sqs from '@aws-cdk/aws-sqs';
+import * as cdk from '@aws-cdk/core';
+import { Service } from '../../service';
+import { Container } from '../container';
+import { ContainerMutatingHook, ServiceExtension } from '../extension-interfaces';
+
+// Keep this import separate from other imports to reduce chance for merge conflicts with v2-main
+// eslint-disable-next-line no-duplicate-imports, import/order
+import { Construct } from '@aws-cdk/core';
+
+/**
+ * An interface that will be implemented by all the resources that can be subscribed to.
+ */
+export interface ISubscribable {
+  readonly subscriptionQueue?: SubscriptionQueue;
+  /**
+   * All classes implementing this interface must also implement the `subscribe()` method
+   */
+  subscribe(extension: QueueExtension): sqs.IQueue;
+}
+
+/**
+ * The settings for the Queue extension.
+ */
+export interface QueueExtensionProps {
+  /**
+   * The list of subscriptions for this service.
+   *
+   * @default none
+   */
+  readonly subscriptions?: ISubscribable[];
+
+  /**
+   * The user-provided default queue for this service.
+   *
+   * @default If the `eventsQueue` is not provided, a default SQS Queue is created for the service.
+   */
+  readonly eventsQueue?: sqs.IQueue;
+
+  /**
+   * The user-provided queue delay fields to configure auto scaling for the default queue.
+   */
+  readonly scaleOnLatency?: QueueAutoScalingOptions;
+}
+
+/**
+ * The topic-specific settings for creating the queue subscriptions.
+ */
+export interface TopicSubscriptionProps {
+  /**
+   * The SNS Topic to subscribe to.
+   */
+  readonly topic: sns.ITopic;
+
+  /**
+   * The user-provided queue to subscribe to the given topic.
+   *
+   * @deprecated use `topicSubscriptionQueue`
+   */
+  readonly queue?: sqs.IQueue;
+
+  /**
+   * The object representing topic-specific queue and corresponding queue delay fields to configure auto scaling.
+   * If not provided, the default `eventsQueue` will subscribe to the given topic.
+   *
+   * @default none
+   */
+  readonly topicSubscriptionQueue?: SubscriptionQueue;
+}
+
+/**
+ * `SubscriptionQueue` represents the subscription queue object which includes the topic-specific queue and its
+ * corresponding auto scaling fields.
+ */
+interface SubscriptionQueue {
+  /**
+   * The user-provided queue to subscribe to the given topic.
+   */
+  readonly queue: sqs.IQueue;
+
+  /**
+   * The user-provided queue delay fields to configure auto scaling for the topic-specific queue.
+   *
+   * @default none
+   */
+  readonly scaleOnLatency?: QueueAutoScalingOptions;
+}
+
+/**
+ * Options for configuring SQS Queue auto scaling.
+ */
+interface QueueAutoScalingOptions {
+  /**
+   * Average amount of time for processing a single message in the queue.
+   */
+  readonly messageProcessingTime: cdk.Duration;
+
+  /**
+   * Acceptable amount of time a message can sit in the queue.
```

Comment:
Should this also include the amount of time to process the message? Is this referring to the "total" latency?

Reply:
Yes, it includes the amount of time required to process the message. I'll update the comment.